src/util/amount.rs

```diff
@@ -0,0 +1,367 @@
+}
+
+#[derive(Default, Clone)]
+pub struct Bitcoin();
+impl Precision for Bitcoin {
+    fn scale() -> i32 {
+        -8
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct Satoshi();
+impl Precision for Satoshi {
+    fn scale() -> i32 {
+        0
+    }
+}
+
+#[derive(Default, Clone)]
+pub struct MilliSatoshi();
+impl Precision for MilliSatoshi {
+    fn scale() -> i32 {
+        3
+    }
+}
+
+type Inner = i64;
+
+fn rescale_inner(inner: Inner, exp: i32) -> Inner {
+    if exp < 0 {
+        inner.checked_div(10i64.pow(-exp as u32)).unwrap()
+    } else {
+        inner.checked_mul(10i64.pow(exp as u32)).unwrap()
+    }
+}
+
+fn pow_round_and_to_inner(v: f64, exp: i32) -> Inner {
+    let amt = v * 10f64.powi(exp);
+    if v < 0.0 {
+        (amt - 0.5) as Inner
+    } else {
+        (amt + 0.5) as Inner
+    }
+}
+
+#[derive(Clone, Copy)]
+pub struct Amount<P = Satoshi>(Inner, P)
+where
+    P: Precision;
+
+impl Amount<Satoshi> {
+    pub fn sat(satoshis: Inner) -> Amount<Satoshi> {
+        Amount(satoshis, Satoshi())
+    }
+}
+
+impl<P: Precision> Amount<P> {
+    pub fn from_inner(amount: Inner) -> Amount<P> {
+        Amount(amount, Default::default())
+    }
+
+    pub fn from_sat(satoshis: i64) -> Amount<P> {
+        assert!(P::scale() >= 0);
+        Amount::from_inner(rescale_inner(satoshis, P::scale()))
+    }
+
+    pub fn from_btc<T: IntoBtc<P>>(btc: T) -> Amount<P> {
+        btc.into_btc()
+    }
+
+    pub fn zero() -> Amount<P> {
+        Amount::from_inner(0)
+    }
+
+    pub fn one() -> Amount<P> {
+        Amount::from_inner(1)
+    }
+
+    pub fn as_inner(&self) -> Inner {
+        self.0
+    }
+
+    pub fn into_inner(self) -> Inner {
+        self.0
+    }
+
+    pub fn float_denomination(&self, denom: Denomination) -> f64 {
+        let exp = -P::scale() - denom as i32;
+        (self.0 as f64) * 10f64.powi(exp)
+    }
+
+    pub fn fmt_denomination(&self, f: &mut fmt::Write, denom: Denomination) -> fmt::Result {
+        write!(f, "{} {}", self.float_denomination(denom.clone()), denom)
+    }
+
+    pub fn to_denominated_string(&self, denom: Denomination) -> String {
+        let mut buf = String::new();
+        self.fmt_denomination(&mut buf, denom).unwrap();
+        buf
+    }
+
+    pub fn rescaled<O: Precision>(&self) -> Amount<O> {
+        let amt = rescale_inner(self.0, O::scale() - P::scale());
+        Amount::from_inner(amt)
+    }
+
+    pub fn into_rescaled<O: Precision>(self) -> Amount<O> {
+        let amt = rescale_inner(self.0, O::scale() - P::scale());
+        Amount::from_inner(amt)
+    }
+}
+
+impl<P: Precision> fmt::Debug for Amount<P> {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Amount({}, presision={})", self.0, P::scale())
+    }
+}
+
+impl fmt::Display for Amount {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        self.fmt_denomination(f, Denomination::Bitcoin)
+    }
+}
+
+impl<P: Precision> ops::Add for Amount<P> {
```

Comment:
Not having a `CheckedAdd` restricts the use of this type to holdings. With enough velocity you could achieve more than 21M BTC in income which would make `Amount` overflow. I don't think this needs a fix, but should be stated in the docs.

Reply:
You're saying we should implement `CheckedAdd` on the type? Fwiw, it shouldn't overflow at 21M but at the max u64 IMO. F.e. a "total received" or so could be perfectly more than 21M.

Reply:
No, I'd just document the maximum somewhere (so people don't have to dig into the code).

Reply:
Oh yeah sure, that's what rust-amount also did. Can add those values again.